epac.1
.\" *****************************************************************************
.\" * Erik's Partial Archive Collator                                           *
.\" * Copyright (C) 2002-2010 Erik Greenwald <erik@smluc.org>                   *
.\" *                                                                           *
.\" * This program takes a directory as an argument, then walks through the     *
.\" * directory looking for duplicate and partially duplicate files. If it      *
.\" * an exact duplicate, it deletes one and creates a hardlink to the other,   *
.\" * minimizing disk usage. If it finds a pair of files where they contain the *
.\" * same data up to the size of the smaller file, it will prompt if you want  *
.\" * to combine them. If you say yes, it will delete the smaller of the files  *
.\" * and hardlink to the larger.                                               *
.\" *                                                                           *
.\" * This program is free software; you can redistribute it and/or modify      *
.\" * it under the terms of the GNU General Public License as published by      *
.\" * the Free Software Foundation; either version 2 of the License, or         *
.\" * (at your option) any later version.                                       *
.\" *                                                                           *
.\" * This program is distributed in the hope that it will be useful,           *
.\" * but WITHOUT ANY WARRANTY; without even the implied warranty of            *
.\" * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the             *
.\" * GNU General Public License for more details.                              *
.\" *                                                                           *
.\" * You should have received a copy of the GNU General Public License         *
.\" * along with this program; if not, write to the Free Software               *
.\" * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA *
.\" *****************************************************************************
.\" epac.1 man page for epac
.\" http://math.missouristate.edu/~erik/software.php?id=92
.\" $Id: epac.1,v 1.7 2010/01/06 01:06:53 erik Exp $
.\"===========================================================================
.Dd August 5, 2002
.Dt EPAC 1
.\"===========================================================================
.Sh NAME
.Nm EPAC
.Nd Reduce (partially) identical files to a single inode.
.\"===========================================================================
.Sh SYNOPSIS
.B epac
.\".Op Fl Crsvh
.Op Fl Cvh
.Op Fl W|--width Ar width
.Op Fl y|n
.Ar dir1
.Op Ar dir2...
.\"===========================================================================
.Sh DESCRIPTION
Erik's Partial Archive Collator takes a directory as an argument, then walks
through it looking for duplicate and partially duplicate files. If it finds
exact duplicates, it deletes one and creates a hard link to the other,
minimizing disk usage. If it finds a pair of files where the larger file
contains the entire contents of the smaller file, it will prompts to combine
them and create a hard link.
.Pp
.\"===========================================================================
.Sh OPTIONS
.Bl -tag -width
.It Op Fl C
Verbose
.\".It Op Fl s
.\"show only, don't collapse
.\".It Op Fl r
.\"Operate recursively
.It Op Fl w|--width Ar width
Force the screen width for the progress bar.
.It Op Fl y
Assume yes to prompted actions (partial matches).
.It Op Fl n
Assume no to prompted actions.
.It Op Fl v
Show version and exit.
.It Op Fl h
Show help and exit.
.El
.\"===========================================================================
.Sh DIAGNOSTICS
.Nm
returns 0 on success and non-zero on failure. The help (-h) and version (-v)
flags return 0. Operational errors return a negative value and interface
(human) errors return a positive value.
.Pp
.Bl -tag -width 4n -offset indent -compact
.It Sy -1
malloc or read error
.It Sy 0
Success, -v, or -h
.It Sy 1
poorly defined option (missing mandatory argument, etc) or -?
.It Sy 2
unknown option
.El
.Pp
.\"===========================================================================
.Sh SEE ALSO
.Xr ln 1 ,
.Xr link 3
.\"===========================================================================
.Sh BUGS
Other than a crappy manpage, 
.Nm
has no known bugs. Please report any to
.An <erik@smluc.org>
.\"===========================================================================
.Sh AUTHOR
.An Erik Greenwald <erik@smluc.org> <erik@math.smsu.edu>
http://math.smsu.edu/~erik/software.php?id=92


ChangeLog
20071205:
	* added "autochoose" (-y|-n) mechanism

20070905:
	* Fix for doc building out of dir.
	* Verify signal before winching.
	* Fix for 64b in comp.c.
	* Improved static testing.
	* Fixed possible buffer overflows.

20051222:
	* Fixed output fd's for help/version messages.
	* Release 0.4.

20051031:
	* Fixes for -Wall -Werror -ansi -pedantic.
	* Expanded some stuff for increased redability.

20051030:
	* Move arg handling to main.c.

20051027:
	* Added a script to perform static testing.

20050105:
	* Autosearch for /usr/local and /sw.

20040725:
	* Error reporting on bad relinking.

20040423:
	* Added -W width flag to force screen width.

20040411:
	* Use src subdir.
	* Print possible matches.
	* Handle different screen widths using $COLUMNS.

20040303:
	* Fixes for MacOSX.3.

20040222:
	* Initial testing stuff.
	* Initial doxygen.

20031227:
	* Release 0.3.
	* RPM stuff.
	* Added linux support.

20031226:
	* Initial manpage.
	* Split display stuff off.
	* Use getopt.

20031225:
	* Move to autoconf.
	* Split list module off.

200300301:
	* Hash table optimization.

20020812:
	* Release 0.2.

20020805:
	* Release 0.1.

20020607:
	* Rewrote 'unify' with improved algorithms, named it 'newunity', still
	  written in Guile Scheme.

20020424:
	* Began project under another name (unify), and written in Scheme 
	  (guile) with a C bit for compute heavy bits.


src/epac.c

/*****************************************************************************
 * Erik's Partial Archive Collator                                           *
 * Copyright (C) 2002-2010 Erik Greenwald <erik@smluc.org>                   *
 *                                                                           *
 * This program takes a directory as an argument, then walks through the     *
 * directory looking for duplicate and partially duplicate files. If it      *
 * an exact duplicate, it deletes one and creates a hardlink to the other,   *
 * minimizing disk usage. If it finds a pair of files where they contain the *
 * same data up to the size of the smaller file, it will prompt if you want  *
 * to combine them. If you say yes, it will delete the smaller of the files  *
 * and hardlink to the larger.                                               *
 *                                                                           *
 * This program is free software; you can redistribute it and/or modify      *
 * it under the terms of the GNU General Public License as published by      *
 * the Free Software Foundation; either version 2 of the License, or         *
 * (at your option) any later version.                                       *
 *                                                                           *
 * This program is distributed in the hope that it will be useful,           *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of            *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the             *
 * GNU General Public License for more details.                              *
 *                                                                           *
 * You should have received a copy of the GNU General Public License         *
 * along with this program; if not, write to the Free Software               *
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA *
 ****************************************************************************/

/*
 * $Id: epac.c,v 1.19 2010/01/06 01:06:53 erik Exp $
 */

#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <unistd.h>
#include <dirent.h>


#include "config.h"

#include "comp.h"
#include "display.h"
#include "epac.h"
#include "list.h"

unsigned int count = 0, inodecount = 0, filecount = 0, possiblematchcount =
    0, at = 0;
unsigned int only_do_savings = 0, do_recursive = 0, verbose = 0, autochoose = -1;

double reclaimed = 0.0;

int
epac (int argc, char **argv)
{
    DIR *d;
    struct dirent *de;
    int i = 0;
    char buf[BUFSIZ];

    /* for each toplevel dir... */
    while (argc--)
    {
	d = opendir (*argv);
	
	if (d == NULL)
	    continue;

	while ((de = readdir (d)) != NULL)
	{
	    static struct stat sb;

	    snprintf (buf, BUFSIZ, "%s/%s", *argv, de->d_name);

	    if (stat (buf, &sb) == -1)
		perror ("stat issue? :");
	    else if (sb.st_mode & S_IFREG && sb.st_size)
		list_add (buf, &sb);

	    if (verbose && !(i % 8))
		printf ("\rFiles: %d\tInodes: %d", filecount, inodecount),
		    fflush (stdout);
	    ++i;
	}
	closedir (d);
	argv++;
    }
    count = inodecount;
    count = (int)((float)count * (float)count / 2.0);
    if (verbose)
	printf ("\nDir read completed, %d inodes, estimating %d scans\n",
	    inodecount, count);

#ifdef DEBUG
    list_print (filelist);
#endif

    if (filelist && filelist->next)
	compagainst (filelist);
    else
	printf ("uh?\n");

    if (verbose)
    {
	showstatus (1.0);
	printf ("\n%d possiblematch calls, %.2f%% scans\n", possiblematchcount,
	    100.0 * possiblematchcount / count);
	printf ("%.0f bytes (%0.2f k, %02.f m) recovered\n", reclaimed,
	    reclaimed / 1024.0, reclaimed / (1024.0 * 1024.0));
    }

    return 0;
}


/**
 */
void
epac_handle_match (struct filegroup_s *a, struct filegroup_s *b)
{
    if (a->size == b->size)
	b = combine (a, b);
    else
    {
	char c[BUFSIZ];

	printf
	    ("\n====================================================================\
	      \n(%d)\t\"%s\"\n(%d)\t\"%s\"\
	      \n====================================================================\n",
	     a->size, a->files->filename, b->size, b->files->filename);
	if (autochoose == 1)
	{
	    b = combine (a, b);
	    return;
	}
	if (autochoose == 0)
	    return;
	while (1)
	{
	    printf ("\aCombine? y/n > ");
	    scanf ("%1s", c);
	    switch (tolower (*c))
	    {
	    case 'y':
		b = combine (a, b);
		return;
	    case 'n':
		return;
	    default:
		printf ("Unknown option \"%c\"\n", *c);
	    }
	}
    }
}
